src/hotspot/cpu/riscv/stubGenerator_riscv.cpp

```diff
@@ -3661,6 +3661,915 @@ class StubGenerator: public StubCodeGenerator {
+    //
+    // VECTORS
+    //  v10 - v13 (1024-bits / 4*256 bits / 4*4*64 bits), hold the message
+    //             schedule words (Wt). They start with the message block
+    //             content (W0 to W15), then further words in the message
+    //             schedule generated via vsha2ms from previous Wt.
+    //   Initially:
+    //     v10 = W[  3:0] = { W3,  W2,  W1,  W0}
+    //     v11 = W[  7:4] = { W7,  W6,  W5,  W4}
+    //     v12 = W[ 11:8] = {W11, W10,  W9,  W8}
+    //     v13 = W[15:12] = {W15, W14, W13, W12}
+    //
+    //  v16 - v17 hold the working state variables (a, b, ..., h)
+    //    v16 = {f[t],e[t],b[t],a[t]}
+    //    v17 = {h[t],g[t],d[t],c[t]}
+    //   Initially:
+    //    v16 = {H5i-1, H4i-1, H1i-1 , H0i-1}
+    //    v17 = {H7i-i, H6i-1, H3i-1 , H2i-1}
+    //
+    //  v0 = masks for vrgather/vmerge. Single value during the 16 rounds.
+    //
+    //  v14 = temporary, Wt+Kt
+    //  v15 = temporary, Kt
+    //
+    //  v18/v19 = temporaries, in the epilogue, to re-arrange
+    //            and byte-swap v16/v17
+    //
+    //  v26/v27 = hold the initial values of the hash, byte-swapped.
+    //
+    //  v30/v31 = used to generate masks, vrgather indices.
+    //
+    // During most of the function the vector state is configured so that each
+    // vector is interpreted as containing four 64 bits (e64) elements (256 bits).
+
+    // Set vectors as 4 * 64
+    //
+    // e64: vector of 64b/8B elements
+    // m1: LMUL=1
+    // ta: tail agnostic (don't care about those lanes)
+    // ma: mask agnostic (don't care about those lanes)
+    // x0 is not written, we known the number of vector elements, 2.
+    __ vsetivli(x0, 4, Assembler::e64, Assembler::m1, Assembler::ma, Assembler::ta);
+
+    // Load H[0..8] to produce
+    //  v16 = {a,b,e,f}
+    //  v17 = {c,d,g,h}
+    __ vle64_v(v16, state);                          // v16 = {d,c,b,a}
+    __ addi(state, state, 32);
+    __ vle64_v(v17, state);                          // v17 = {h,g,f,e}
+
+    __ vid_v(v30);                                   // v30 = {3,2,1,0}
+    __ vxor_vi(v30, v30, 0x3);                       // v30 = {0,1,2,3}
+    __ vrgather_vv(v26, v16, v30);                   // v26 = {a,b,c,d}
+    __ vrgather_vv(v27, v17, v30);                   // v27 = {e,f,g,h}
+    __ vmsgeu_vi(v0, v30, 2);                        // v0  = {f,f,t,t}
+    // Copy elements [3..2] of v26 ({d,c}) into elements [3..2] of v17.
+    __ vslideup_vi(v17, v26, 2);                     // v17 = {c,d,_,_}
+    // Merge elements [1..0] of v27 ({g,h}) into elements [1..0] of v17
+    __ vmerge_vvm(v17, v17, v27);                    // v17 = {c,d,g,h}
+    // Copy elements [1..0] of v27 ({f,e}) into elements [1..0] of v16.
+    __ vslidedown_vi(v16, v27, 2);                   // v16 = {_,_,e,f}
+    // Merge elements [3..2] of v26 ({a,b}) into elements [3..2] of v16
+    __ vmerge_vvm(v16, v26, v16);                    // v16 = {a,b,e,f}
```

Comment:
Simlar here. Can we make use of index-load and index-store to simplify the code for 512 case too?

Reply:
Not yet addressed.

Reply:
Fixed